Improvement + Fix: Pests Rework

ItsEmpa opened the pull request:
## What
Fixes a variety of issues with pests, and adds support for getting more information from other sources, such as the tablist widget, offline pest messages, specific pests in plots from scoreboard, etc.

<details>
<summary>TODO</summary>

Note: no idea how to implement these.
- [ ] Add support for offline pests spawn messages in modified pest spawn message.
- [ ] Fix pests not being properly removed when clearing pests from another plot while being in an infected plot.

</details>
<details>
<summary>Images</summary>


</details>

## Changelog Improvements
+ Added support for using the tab list pest widget to know where pests are. - Empa
+ Use more information about pests from the scoreboard when possible. - Empa
+ Added support for offline pest messages. - Empa

## Changelog Fixes
+ Fixed pest overlays not showing if the amount in each one was unknown. - Empa
+ Fixed pests spawning in the Barn plot not being detected. - Empa

## Changelog Technical Details
+ Added "unknownAmount" to PestSpawnEvent. - Empa
+ Added PestUpdateEvent. - Empa
+ Moved internal pest logic to PestAPI. - Empa

ItsEmpa committed:
pests rework

ItsEmpa committed:
formatting

hannibal002 commented:
```
SkyHanni 0.24.Beta.17: Caught an IndexOutOfBoundsException at TabListUpdateEvent in PestFinder: Index: 0, Size: 0
 
Caused by java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:657)
	at java.util.ArrayList.get(ArrayList.java:433)
	at SH.features.garden.pests.PestFinder.fixPests(PestFinder.kt:149)
	at SH.features.garden.pests.PestFinder.onTablistUpdate(PestFinder.kt:283)
	at SH.utils.TabListData.onTick(TabListData.kt:112)
	at SH.data.MinecraftData.onTick(MinecraftData.kt:74)
	at FML.common.eventhandler.EventBus.post(EventBus.java:140)
```

ItsEmpa committed:
formatting and error

ItsEmpa commented:
forgot to change the ones in the world, also think i fixed that error

ItsEmpa committed:
remove unused comment

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/pests/PestFinder.kt:
```diff
@@ -57,29 +92,41 @@ class PestFinder {
-            for (line in item.getLore()) {
-                pestInventoryPattern.matchMatcher(line) {
-                    plot.pests = group("amount").formatNumber().toInt()
+            if (!plot.isBarn()) {
```
early return

ItsEmpa replied:
this cant be an early return i think, since the reason this check is here is to not accidentally set the amount of pests in the barn plot as 0, as hypixel doesnt show the amount of pests in it in the plots menu

hannibal002 replied:
continue, not return. just switch the if statement. intellij should suggest you the correct action already

ItsEmpa replied:
weird intellij wasnt suggesting anything

hannibal002 reviewed (changes requested):
fixPests() always calls update(), but fixPests only gets called immediately before update gets called.
also maybe just fixPests at start of update?

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/pests/PestFinder.kt:
```diff
@@ -139,54 +211,96 @@ class PestFinder {
+            noPestsInScoreboardPattern.matchMatcher(line) {
+                if (PestAPI.scoreboardPests != 0) {
+                    PestAPI.scoreboardPests = 0
+                    GardenPlotAPI.plots.forEach {
+                        it.pests = 0
+                        it.isPestCountInaccurate = false
+                    }
```
put those lines into a function, and use it two times. put the update call at the end in the function. call it something like resetAllPests idk

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/pests/PestFinder.kt:
```diff
@@ -139,54 +211,96 @@ class PestFinder {
     fun onScoreboardChange(event: ScoreboardChangeEvent) {
```
add comments above the different matchMatcher calls. its not intuitive clear what is happening at every single spot

hannibal002 reviewed (changes requested):
maybe move all the intnernal pest detection and store logic into a pest api class?

hannibal002 commented:
This feature works fine to me btw, just code cleanup ideas

ItsEmpa committed:
cleanup

ItsEmpa committed:
continue instead of if

ItsEmpa commented:
> maybe move all the intnernal pest detection and store logic into a pest api class?

will do yes

ItsEmpa committed:
move all pest logic to pestapi

ItsEmpa commented:
Moved all the internal pest logic into PestAPI

hannibal002 committed:
Merge branch 'beta' into fork/pests-rework

hannibal002 reviewed (changes requested):
idk why this happened

<details>
<summary>Details</summary>

```
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: Caused by: java.lang.ExceptionInInitializerError
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	at at.hannibal2.skyhanni.features.garden.pests.SprayFeatures.getConfig(SprayFeatures.kt:23)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	at at.hannibal2.skyhanni.features.garden.pests.SprayFeatures.isEnabled(SprayFeatures.kt:78)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	at at.hannibal2.skyhanni.features.garden.pests.SprayFeatures.onRenderOverlay(SprayFeatures.kt:57)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_1164_SprayFeatures_onRenderOverlay_GuiOverlayRenderEvent.invoke(.dynamic)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:49)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	at at.hannibal2.skyhanni.events.LorenzEvent.postAndCatchAndBlock(LorenzEvent.kt:43)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:634]: 	... 20 more
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: Caused by: java.lang.RuntimeException: Illegal late initialization of repo pattern. Repo pattern needs to be created during pre-initialization.
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: 	at at.hannibal2.skyhanni.utils.repopatterns.RepoPatternManager.crash(RepoPatternManager.kt:55)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: 	at at.hannibal2.skyhanni.utils.repopatterns.RepoPatternManager.of(RepoPatternManager.kt:151)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: 	at at.hannibal2.skyhanni.utils.repopatterns.RepoPattern$Companion.pattern(RepoPattern.kt:94)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: 	at at.hannibal2.skyhanni.utils.repopatterns.RepoPatternGroup.pattern(RepoPatternGroup.kt:18)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: 	at at.hannibal2.skyhanni.features.garden.pests.PestAPI.<clinit>(PestAPI.kt:45)
[20:59:44] [main/INFO] (STDERR) [java.lang.Throwable:printStackTrace:643]: 	... 26 more
```

</details>

hannibal002 committed:
fix

ItsEmpa commented:
forgot to add 2 things to the changelog yesterday, added them now, theyre just

+ Created PestUpdateEvent. - Empa
+ Moved internal pest logic to PestAPI. - Empa
